### PR TITLE
ED-1231 - Add oci oss support for Build/DataPipeline/AnalyticsCore

### PR DESCRIPTION
**Pull request by ddevadat**
updated cloud-store-sdk to 1.4.6 along with associated class file changes

### Type of change

Please choose appropriate options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### How Has This Been Tested?

Please describe the tests that you ran to verify your changes in the below checkboxes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [X] Executed jenkins build job Build/DataPipeline/AnalyticsCore

### CODE REVIEW

File: analytics-core/src/main/scala/org/ekstep/analytics/framework/conf/AppConf.scala [hunk 1]
```diff
@@ -12,9 +12,13 @@ object AppConf {
     lazy val conf = defaultConf.withFallback(envConf);
 
     def getConfig(key: String): String = {
-        if (conf.hasPath(key))
+        val keyValue = if (conf.hasPath(key)){
```
Comment (anandp504) [inline analytics-core/src/main/scala/org/ekstep/analytics/framework/conf/AppConf.scala:15]: Can this be handled as an Option with getOrElse? Or if it is a scala map, we can directly do a getOrElse if I am right.

File: analytics-core/src/main/scala/org/ekstep/analytics/framework/conf/AppConf.scala [hunk 2]
```diff
@@ -12,9 +12,13 @@ object AppConf {
...
             conf.getString(key);
-        else "";
+        } else {
+            ""
+        }
+        println(s"The value for $key is $keyValue")
```
Comment (anandp504) [inline analytics-core/src/main/scala/org/ekstep/analytics/framework/conf/AppConf.scala:20]: Please remove the console println statements.

File: analytics-core/src/main/scala/org/ekstep/analytics/framework/util/MergeUtil.scala [hunk 3]
```diff
@@ -74,8 +74,29 @@ class MergeUtil {
           columnOrder = columnOrder.filter(col=> reportDfColumns.contains(col))
           MergeResult(mergeReport(rollupCol,rollupFormat,deltaDF, reportDF, mergeConfig, mergeConfig.merge.dims), reportDF,
             StorageConfig(storageType, postContainer, path.getParent.getName,Option(storageKey),Option(storageSecret)))
+        case "s3" =>
+          var deltaDF = fetchOSSFile(filePaths("deltaPath"),
```
Comment (anandp504) [inline analytics-core/src/main/scala/org/ekstep/analytics/framework/util/MergeUtil.scala:78]: What is the use of this deltaDF?

File: analytics-core/src/main/scala/org/ekstep/analytics/framework/util/MergeUtil.scala [hunk 4]
```diff
@@ -74,8 +74,29 @@ class MergeUtil {
...
+            mergeConfig.deltaFileAccess.getOrElse(true), mergeConfig.container)
+            deltaDF = (if(deltaDF.columns.contains("Date")) deltaDF.withColumn(rollupCol,
+              date_format(col("Date"), rollupFormat)) else deltaDF).dropDuplicates()
+          val reportFile = fetchOSSFile(filePaths("reportPath"), mergeConfig.reportFileAccess.getOrElse(true),
+            postContainer)
+          val reportDF = if (null == reportFile) {
+            sqlContext.createDataFrame(sc.emptyRDD[Row], if (null == reportSchema) {
```
Comment (anandp504) [inline analytics-core/src/main/scala/org/ekstep/analytics/framework/util/MergeUtil.scala:85]: reportSchema is always initialised to null. What is the use of this if-else statement?

File: analytics-core/src/test/resources/application.conf [hunk 5]
```diff
@@ -60,4 +60,17 @@ log.appender.kafka.topic="telemetry.log"
...
 druid.report.default.storage="azure"
 druid.report.date.format="yyyy-MM-dd"
-druid.report.default.container="report-verification"
\ No newline at end of file
+druid.report.default.container="report-verification"
+
+
+#Monitor Jobs
+
+monitor {
```
Comment (anandp504) [inline analytics-core/src/test/resources/application.conf:68]: Do we need this flink-alerts notification?

**Review by anandp504 — CHANGES_REQUESTED** [review]
@ddevadat Please resolve the PR comments.

File: analytics-core/src/main/scala/org/ekstep/analytics/framework/util/MergeUtil.scala (hunk 3 above)
Comment (ddevadat) [reply analytics-core/src/main/scala/org/ekstep/analytics/framework/util/MergeUtil.scala:78]: This block of code for case s3 , is a direct replica for the case of azure, with the relevant difference like 
var deltaDF = fetchBlobFile(filePaths("deltaPath") changed to var deltaDF = fetchOSSFile(filePaths("deltaPath")

here s3 refers to the deep storage type used by druid. There was no implementation when the deep storage used by druid is of type s3.
This block of change was to fill the gap for s3 storage type with the existing business logic of type azure.

File: analytics-core/src/main/scala/org/ekstep/analytics/framework/util/MergeUtil.scala (hunk 4 above)
Comment (ddevadat) [reply analytics-core/src/main/scala/org/ekstep/analytics/framework/util/MergeUtil.scala:85]: Same as the previous comment, this is just the replica of the case for azure ( line 64 ). We just replicated the existing working logic of azure to make it work for s3 type of storage

File: analytics-core/src/main/scala/org/ekstep/analytics/framework/conf/AppConf.scala (hunk 1 above)
Comment (ddevadat) [reply analytics-core/src/main/scala/org/ekstep/analytics/framework/conf/AppConf.scala:15]: this was done for some debugging purpose. reverted to the original form

File: analytics-core/src/main/scala/org/ekstep/analytics/framework/util/MergeUtil.scala [hunk 6]
```diff
@@ -181,6 +202,20 @@ class MergeUtil {
     else null
   }
 
+  def fetchOSSFile(filePath: String, isPrivate: Boolean, container: String)(implicit sqlContext: SQLContext, fc: FrameworkContext): DataFrame = {
+
+    val storageService = if (isPrivate)
+      fc.getStorageService("oci", "azure_storage_key", "azure_storage_secret")
```
[the remark this answers is not in the document]
Comment (anandp504) [reply analytics-core/src/main/scala/org/ekstep/analytics/framework/util/MergeUtil.scala:208]: @ddevadat I would prefer to use the AWS variables as you had suggested to make things more understandable. With using Azure variables, it makes it difficult to understand.

[the remark this answers is not in the document]
Comment (ddevadat) [reply analytics-core/src/main/scala/org/ekstep/analytics/framework/util/MergeUtil.scala:208]: changed to use aws_storage_key & aws_storage_secret

**Pull request merged**